pkg/migrations/sql/preinstall/006-set_up_ha.sql

```diff
@@ -0,0 +1,143 @@
+    ON SCHEMA_CATALOG.ha_leases
+    FOR EACH ROW
+EXECUTE PROCEDURE SCHEMA_CATALOG.ha_leases_audit_fn();
+
+-- default values for lease
+INSERT INTO SCHEMA_CATALOG.default(key, value)
+VALUES ('ha_lease_timeout', '1m'),
+       ('ha_lease_refresh', '10s')
+ON CONFLICT (key) DO UPDATE SET value = EXCLUDED.value;
+
+-- ha api functions
+CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.update_lease(cluster TEXT, writer TEXT, min_time TIMESTAMPTZ,
+                                                       max_time TIMESTAMPTZ, lease_timeout INTERVAL DEFAULT '1m',
+                                                       lease_refresh INTERVAL DEFAULT '10s') RETURNS ha_leases
+AS
+$func$
+DECLARE
+    leader            TEXT;
+    lease_start       TIMESTAMPTZ;
+    lease_until       TIMESTAMPTZ;
+    new_lease_timeout TIMESTAMPTZ;
+    lease_state       ha_leases%ROWTYPE;
+BEGIN
+
+    -- find latest leader and their lease time range;
+    SELECT h.leader_name, h.lease_start, h.lease_until
+    INTO leader, lease_start, lease_until
+    FROM SCHEMA_CATALOG.ha_leases as h
+    WHERE cluster_name = cluster;
+
+    --only happens on very first call;
+    IF NOT FOUND THEN
+        -- no leader yet for cluster insert;
+        INSERT INTO SCHEMA_CATALOG.ha_leases
+        VALUES (cluster, writer, min_time, max_time + lease_timeout)
+        ON CONFLICT DO NOTHING;
+        -- needed due to on-conflict clause;
+        SELECT h.leader_name, h.lease_start, h.lease_until
+        INTO leader, lease_start, lease_until
+        FROM SCHEMA_CATALOG.ha_leases as h
+        WHERE cluster_name = cluster;
+    END IF;
+
+    IF leader <> writer THEN
+        RAISE EXCEPTION 'LEADER_HAS_CHANGED' USING ERRCODE = 'PS010';
+    END IF;
+
+    new_lease_timeout = max_time + lease_timeout;
+    IF new_lease_timeout > lease_until + lease_refresh THEN
+        UPDATE SCHEMA_CATALOG.ha_leases h
+        SET lease_until = new_lease_timeout
+        WHERE h.cluster_name = cluster
+          AND h.leader_name = writer
+          AND h.lease_until + lease_refresh < new_lease_timeout;
+        IF NOT FOUND THEN -- concurrent update
+            SELECT h.leader_name, h.lease_start, h.lease_until
+            INTO leader, lease_start, lease_until
+            FROM SCHEMA_CATALOG.ha_leases as h
+            WHERE cluster_name = cluster;
+            IF leader <> writer OR lease_until < max_time
+            THEN
+                RAISE EXCEPTION 'LEADER_HAS_CHANGED' USING ERRCODE = 'PS010';
+            END IF;
+        END IF;
+    END IF;
+    SELECT * INTO lease_state FROM ha_leases WHERE cluster_name = cluster;
+    RETURN lease_state;
+END;
+$func$ LANGUAGE plpgsql VOLATILE;
+
+CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.try_change_leader(cluster TEXT, new_leader TEXT,
+                                                            max_time TIMESTAMPTZ,
+                                                            lease_timeout INTERVAL DEFAULT '1m') RETURNS ha_leases
```

Comment:
shouldn't lease_timeout be set from the default setting in the db as opposed to a parameter?